util/network-devp2p/src/enr.rs

```diff
@@ -0,0 +1,434 @@
+// Copyright 2015-2019 Parity Technologies (UK) Ltd.
+// This file is part of Parity Ethereum.
+
+// Parity Ethereum is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity Ethereum is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Ethereum.  If not, see <http://www.gnu.org/licenses/>.
+
+//! This module implements Ethereum Node Records as defined in EIP-778. A node record holds
+/// arbitrary information about a node on the p2p network.
+
+/// Node information is stored in key/value pairs.
+
+/// Node records must be signed before transmitting them to another node.
+
+/// Decoding a record doesn't check its signature.
+/// When receiving records from untrusted peer, A node must verify two things:
+/// - identity scheme.
+/// - the signature is valid according to the declared scheme.
+
+/// Whenever you create or modify a record, use a signing function provided by
+/// the identity scheme to add the signature.
+
+use std::collections::HashMap;
+use rlp::{encode, Encodable, RlpStream};
+use std::net::IpAddr;
+use log::error;
+
+/// Maximum encoded size of a node record in bytes.
+pub const ENR_MAX_SIZE: usize = 300;
+
+/// A node record entry.
+/// A new node record entry should implements this trait.
+pub trait Entry {
+    fn enr_key(&self) -> String;
+}
+
+/// Provide custom verify method for each scheme.
+pub trait Scheme {
+    fn verify(&self, r: &Record, sig: &Vec<u8>) -> bool;
+}
+
+/// Holes scheme for some type.
+pub struct SchemeMap<T: Scheme>(HashMap<String, T>);
+
+impl<T: Scheme> SchemeMap<T> {
```

Comment:
Can you explain the relation between `SchemeMap` and `Record`?

Reply:
Current ENR scheme is only "v4".
But may be added later.
`SchemeMap` contains many schemes like "v4" and others.
`Record` represents one version of scheme.
This implementation supports many version of schemes.

Reply:
*identity scheme* is defined as a predefined `id` key in the spec
https://eips.ethereum.org/EIPS/eip-778
I'm not sure we need `SchemeMap` here as well
also see https://github.com/sigp/rust-libp2p/blob/discv5/misc/enr/src/lib.rs